src/nnet3/nnet-cudnn-simple-component.cc

```diff
@@ -0,0 +1,798 @@
+  int32 strides[kConvolutionDimension_];
+  ExpectToken(is, binary, "<FilterXStride>");
+  ReadBasicType(is, binary, &strides[0]);
+  ExpectToken(is, binary, "<FilterYStride>");
+  ReadBasicType(is, binary, &strides[1]);
+  ExpectToken(is, binary, "<FilterZStride>");
+  ReadBasicType(is, binary, &strides[2]);
+  int32 upscales[kConvolutionDimension_];
+  ExpectToken(is, binary, "<FilterXUpscale>");
+  ReadBasicType(is, binary, &upscales[0]);
+  ExpectToken(is, binary, "<FilterYUpscale>");
+  ReadBasicType(is, binary, &upscales[1]);
+  ExpectToken(is, binary, "<FilterZUpscale>");
+  ReadBasicType(is, binary, &upscales[2]);
+  ExpectToken(is, binary, "<InputVectorization>");
+  int32 input_vectorization;
+  ReadBasicType(is, binary, &input_vectorization);
+  input_vectorization_ = static_cast<TensorVectorizationType>(input_vectorization);
+  ExpectToken(is, binary, "<FilterParams>");
+  filter_params_.Read(is, binary);
+  ExpectToken(is, binary, "<BiasParams>");
+  bias_params_.Read(is, binary);
+  ExpectToken(is, binary, "<IsGradient>");
+  ReadBasicType(is, binary, &is_gradient_);
+  ExpectToken(is, binary, "</CuDNN3DConvolutionComponent>");
+
+  CUDNN_SAFE_CALL(cudnnSetFilterNdDescriptor(filter_desc_,
+                                             cudnn::GetDataType(),
+                                             kConvolutionDimension_,
+                                             filter_dims
+                                             ));
+  // N x C x D x H x W
+  // TODO: torch uses {1, num_filters, 1, 1}
+  // Understand why. Is this a mistake?
+  int32 bias_dims[] = {1, num_filters_, 1, 1};
+  int32 bias_strides[] = {num_filters_, 1, 1, 1};
+  CUDNN_SAFE_CALL(cudnnSetTensorNdDescriptor(bias_desc_,
+                                             cudnn::GetDataType(),
+                                             kConvolutionDimension_ + 1,
+                                             bias_dims,
+                                             bias_strides
+                                             ));
+  CUDNN_SAFE_CALL(cudnnSetConvolutionNdDescriptor(conv_desc_,
+                                                  kConvolutionDimension_,
+                                                  padding,
+                                                  strides,
+                                                  upscales,
+                                                  CUDNN_CROSS_CORRELATION,
+                                                  cudnn::GetDataType()
+                                                  ));
+}
+
+void CuDNN3DConvolutionComponent::Write(std::ostream &os, bool binary) const {
+  WriteUpdatableCommon(os, binary);  // write opening tag and learning rate.
+  WriteToken(os, binary, "<InputXDim>");
+  WriteBasicType(os, binary, input_x_dim_);
+  WriteToken(os, binary, "<InputYDim>");
+  WriteBasicType(os, binary, input_y_dim_);
+  WriteToken(os, binary, "<InputZDim>");
+  WriteBasicType(os, binary, input_z_dim_);
```

Comment:
It is actually a good idea to have the interface looked close to the original ConvolutionComponent. It is better to rename x-stride to x-step at the interface.